### PR TITLE
Added options enable-testing and BUILD_SHARED_LIBS

### PR DESCRIPTION
**Pull request by bilke**
I had problems using the `enable-static`-option. Either it did not configure (with the Ninja generator) or it did not build. CMake already provides functionality to switch between static and shared libs with the `BUILD_SHARED_LIBS`-option. If you agree with these changes I think we should remove the `enable-static`-option too.

Also I added a switch to disable testing (`-Denable-testing=OFF`).

### CODE REVIEW

**Comment by thelfer** [comment]
@bilke Thanks for this proposal. I must admit that I have not used the `enable-static` option for a while. Anything conforming to the `cmake` standard is welcomed. Let me make some tests before merging. I also will need to backport this in `rliv-4.0` and the `master` branch.

**Comment by thelfer** [comment]
@bilke I now remember why the `enable-static` option was introduced. It was meant to compile both the shared **and** the static libraries. I probably not used `BUILD_SHARED_LIBS` option because I shall have considered that it did not make sense to compile `MFront` statically. Obviously, I was wrong because you seem to need it, although I am not sure to understand why.

## Keeping the the `enable-static` option

However, I would keep the `enable-static` option. For consistency, I would check if that this option is `OFF` is `BUILD_SHARED_LIBS` is off (otherwise, the static library would be built twice). Could you do that ?

Also, there is a trouble. Those lines must be put before any `add_subdirectory` command in the main `CMakeLists.txt`:

```cmake
option(enable-testing "enable tests" ON)
if(enable-testing)
  add_subdirectory(tests)
endif()
```

## Modules in `mfront/tests`

However, if you do, then at lot of errors are triggered in the `mfront/tests/models/interfaces/src/CMakeLists.txt`  and `mfront/tests/properties/interfaces/src/CMakeLists.txt` directories because some libraries are declared as modules. In those directories, the line

```cmake
if(NOT WIN32)
```

shall be changed in:

```cmake
if((NOT WIN32) AND BUILD_SHARED_LIBS)
```

Would you do that ?

## Tests

Then, come the tests, which is way more compilated...

### The  `mfront_behaviour_check_library` and   `mfront_behaviour_brick check_library`  functions

In pratice, the functions `mfront_behaviour_check_library` and   `mfront_behaviour_brick check_library` in `cmake/modules/tfel.cmake` shall be modified as follows:

```cmake
  if(BUILD_SHARED_LIBS)
    add_library(${lib} MODULE EXCLUDE_FROM_ALL
      ${${lib}_SOURCES}
      ${${lib}_ADDITIONAL_SOURCES})
  else(BUILD_SHARED_LIBS)
    add_library(${lib} EXCLUDE_FROM_ALL
      ${${lib}_SOURCES}
      ${${lib}_ADDITIONAL_SOURCES})
  endif(BUILD_SHARED_LIBS)
```

### The `test_generic` function and similar in **all** interfaces

`MTest` tests shall be desactivated as follows for **all** interfaces.

Let us take the example of the `generic` interface. The easiest way to do this is to rename the functions `test_generic` and `test_generic_fs` in `test_generic_internal` and  `test_generic_fs_internal` respectively. Those functions can then be implemented as follows:

```cmake
function(test_generic)
  if(BUILD_SHARED_LIBS)
    test_generic_internal(${ARGN})
  endif(BUILD_SHARED_LIBS)
endfunction(test_generic)

function(test_generic_fs)
  if(BUILD_SHARED_LIBS)
    test_generic_fs_internal(${ARGN})
  endif(BUILD_SHARED_LIBS)
endfunction(test_generic_fs)
```

## The `check_all.sh` script

Could you add at least a case testing the case `BUILD_SHARED_LIBS=OFF` ? This shall not be executed in case of fast testing (i.e. if `fcheck` is yes).

Let me know when this script works !

**Comment by bilke** [comment]
Thanks for your comments!

I will answer more thoroughly tomorrow..

For today I fixed the `option(enable-testing "enable tests" ON)` declaration.

I also run the `check`-target locally (macOS) both with `BUILD_SHARED_LIBS=ON` and `BUILD_SHARED_LIBS=OFF` and it run successfully. I will check the module stuff tomorrow...

**Comment by bilke** [comment]
In my opinion the modules should be fine as is because they are always build as shared libraries. For me the following works:

```
mkdir build-shared
cd build-shared
cmake .. -DCMAKE_BUILD_TYPE=Release -Denable-aster=ON -Denable-abaqus=ON -Denable-calculix=ON -Denable-comsol=ON -Denable-diana-fea=ON -Denable-ansys=ON -Denable-europlexus=ON -Denable-zmat=ON -Denable-cyrano=ON -Denable-lsdyna=ON -DBUILD_SHARED_LIBS=ON -G Ninja
ninja check
cd ..

mkdir build-static
cmake .. -DCMAKE_BUILD_TYPE=Release -Denable-aster=ON -Denable-abaqus=OFF -Denable-calculix=ON -Denable-comsol=ON -Denable-diana-fea=ON -Denable-ansys=ON -Denable-europlexus=ON -Denable-zmat=ON -Denable-cyrano=ON -Denable-lsdyna=ON -DBUILD_SHARED_LIBS=OFF -G Ninja
ninja check
```

Please note that on the static build I needed to disable the abaqus interface (`-Denable-abaqus=OFF`), the `check`-target failed.

I have not added a test to `check_all.sh` yet as I was not able to run it:

```
$ ./check_all.sh
~/code/ogs/tfel/build-check ~/code/ogs/tfel
~/code/ogs/tfel/build-check/autotools ~/code/ogs/tfel/build-check ~/code/ogs/tfel
~/code/ogs/tfel/build-check/autotools/build-autotools ~/code/ogs/tfel/build-check/autotools ~/code/ogs/tfel/build-check ~/code/ogs/tfel
./check_all.sh: line 124: /home/bilke/code/ogs/tfel/configure: No such file or directory
```

**Comment by thelfer** [comment]
> modules should be fine as is because they are always build as shared libraries

On my Linux box, it fails as follows:

```bash
[ 91%] Linking CXX shared module libLicosMFrontModel.so
/usr/bin/ld: ../../../../src/libTFELMFront.a(LibraryDescription.o): relocation R_X86_64_PC32 against symbole `_ZTISt13runtime_error@@GLIBCXX_3.4' can not be used when making un objet partagé; recompilez avec -fPIC
`_ZTISt13runtime_error@@GLIBCXX_3.4' can not be used when making un objet partagé; recompilez avec -fPIC
.....
/usr/bin/ld: ../../../../src/libTFELMFront.a(DSLUtilities.o): relocation R_X86_64_PC32 against symbole /usr/bin/ld : échec de l'édition de liens finale : nonrepresentable section on output
```
Clearly, `Linux` and `Mac Os` are very different beasts.

> I have not added a test to check_all.sh yet as I was not able to run it:

Argh, you shall first call the `bootstrap.sh` script to call all autotools magic commands... Never tried on `Mac Os` BTW :)

**Comment by thelfer** [comment]
> On my Linux box, it fails as follows:

My bad. It works if I add (as you did):

```cmake
if(NOT BUILD_SHARED_LIBS)
  set(CMAKE_POSITION_INDEPENDENT_CODE ON)
endif()
```

Goods news. With this option, no need to modify any of the `mfront_behaviour_check_library`, `mfront_behaviour_brick`, `test_generic` and `test_generic_fs` functions !

**Comment by thelfer** [comment]
`python` bindings works also as expected. Great !

File: mfront/tests/models/interfaces/src/CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@ if(NOT WIN32)
     PRIVATE TFELMFront
     PRIVATE MFrontLogStream
     PRIVATE TFELSystem)
-  add_library(GerminalMFrontModel SHARED
+  add_library(GerminalMFrontModel
```
Comment (thelfer) [inline mfront/tests/models/interfaces/src/CMakeLists.txt:9]: Shall be a module here

**Comment by thelfer** [comment]
I think that we are close to merging. I just made two small comments.
Would you also prepare an update for the release notes (file `release-notes-3.4.4.md` in `docs/web`)  in the **`master`** branch?
(The documentation is currently kept up-to-date only in the `master` branch. I really shall put it in a separate project...)

**Pull request merged**